pyproject.toml

```diff
@@ -38,11 +38,11 @@ numpy = [
 ]
 six = "^1.16"
 # This functionality was merged into python stdlib beginning in 3.8
-importlib_metadata = {version="^4.10", python="~3.7"}
+importlib_metadata = {version="4.2", python="~3.7"}
 # Documentation, must be in main dependencies (but optional) list for
 # readthedocs integration.
 # See https://github.com/readthedocs/readthedocs.org/issues/4912.
-Sphinx = {version="^4.4", optional=true}
+Sphinx = {version="^4.3", optional=true}
 sphinx_rtd_theme = {version="^1.0", optional=true}
```

Comment:
Why did you downgrade these?

Reply:
This question is still relevant

Reply:
To answer my own question, if we don't downgrade these, `poetry lock` fails with:

```
  SolverProblemError

  Because no versions of flake8 match >5.0,<5.0.1 || >5.0.1,<5.0.2 || >5.0.2,<5.0.3 || >5.0.3,<5.0.4 || >5.0.4,<6.0
   and flake8 (5.0.0) depends on importlib-metadata (<4.3), flake8 (>=5.0,<5.0.1 || >5.0.1,<5.0.2 || >5.0.2,<5.0.3 || >5.0.3,<5.0.4 || >5.0.4,<6.0) requires importlib-metadata (<4.3).
  And because flake8 (5.0.1) depends on importlib-metadata (<4.3)
   and flake8 (5.0.2) depends on importlib-metadata (<4.3), flake8 (>=5.0,<5.0.3 || >5.0.3,<5.0.4 || >5.0.4,<6.0) requires importlib-metadata (<4.3).
  And because flake8 (5.0.3) depends on importlib-metadata (<4.3)
   and flake8 (5.0.4) depends on importlib-metadata (>=1.1.0,<4.3), flake8 (>=5.0,<6.0) requires importlib-metadata (<4.3).
  Because no versions of ni-python-styleguide match >0.4,<0.5
   and ni-python-styleguide (0.4.0) depends on flake8 (>=5.0,<6.0), ni-python-styleguide (>=0.4,<0.5) requires flake8 (>=5.0,<6.0).
  Thus, ni-python-styleguide (>=0.4,<0.5) requires importlib-metadata (<4.3).
  So, because nidaqmx depends on both importlib_metadata (^4.10) and ni-python-styleguide (^0.4), version solving failed.
```

If we do downgrade these, `poetry lock` hangs forever on
```
   2: selecting sphinxcontrib.jquery (2.0.0)
   2: selecting sphinxcontrib.jquery (2.0.0)
   2: selecting sphinxcontrib.jquery (2.0.0)
   2: selecting sphinxcontrib.jquery (2.0.0)
   2: selecting sphinxcontrib.jquery (2.0.0)
   2: selecting sphinxcontrib.jquery (2.0.0)
   2: selecting sphinxcontrib.jquery (2.0.0)
```

Cc: @mshafer-NI

Reply:
Options:
1. keep the downgrade till June (when Python 3.7 is EOL and can officially be dropped)
1. Move sphinx (and related) to a separate env that doesn't participate in the lock (this is what we're doing with a lot of modules internally)
1. Drop 3.7 support early (the conflict on importlib-metada is a <3.8 backport dep)

I would suggest number 1 as these constraints don't affect customers, but dropping 3.7 woulld.

Reply:
Okay, option 1 sounds fine to me.

Regarding `poetry lock` hanging:
- The `sphinxcontrib.jquery` hang seems to be related to upgrading `sphinx_rtd_theme` from 1.0 to 1.1 or 1.2.
- Poetry spends a few minutes downloading multiple versions of `numpy` to figure out that Python 3.7 needs version 1.21.6 but Python 3.8+ can use version 1.24.2. I guess this is acceptable because it finishes in less than 5 minutes.
- 
These are dependencies of nidaqmx-python, not just dev-dependencies, so specifying a version constraint would affect customers. @mshafer-NI How should we handle this?

@Sakthi-SM Did `poetry lock` hang when you ran it? How long did it take?

Reply:
For numpy, you can speed that up by telling poetry (and pip via the published whl) to use an older version for 3.7 

(something like):
```toml
"numpy" = [
  {version=">=1.20,<1.24.2", python="<3.8"},
  {version=">=1.24.2", python="^3.8"},
]
```

Reply:
`numpy` dropped Python 3.7 support in version 1.22, so I think the 1st line should have `version=">=1.20,<1.22", python="<3.8"`.

Reply:
We can add `sphinx_rtd_theme` to dev-dependencies with a stricter version constraint:
```
sphinx_rtd_theme = "1.0"
```

With this and the updated `numpy` constraints, `poetry lock` completes in a reasonable amount of time on my machine.